schema-registry/authorizer/authorization-agent/src/main/java/com/hortonworks/registries/schemaregistry/authorizer/agent/DefaultAuthorizationAgent.java

```diff
@@ -0,0 +1,515 @@
+        });
+
+    }
+
+    ////////////////////////////////////////////////////////////////////////
+
+    public static class AlreadyConfiguredException extends RuntimeException {
+        public AlreadyConfiguredException(String message) {
+            super(message);
+        }
+    }
+
+    private interface EntityFilterFunction<T> {
+        T filter(T elem) throws AuthorizationException, SchemaNotFoundException;
+    }
+
+    private interface EntityToAuthorizerResourceMapFunc<T> {
+        Authorizer.Resource map(T elem) throws SchemaNotFoundException;
+    }
+
+    private <T> Collection<T> authorizeGetEntities(UserAndGroups userAndGroups,
+                                                   Collection<T> entities,
+                                                   EntityToAuthorizerResourceMapFunc<T> mapFunc)
+            throws SchemaNotFoundException {
+        return removeUnauthorizedAndNullEntities(entities, elem ->
+                authorizer.authorize(mapFunc.map(elem), AccessType.READ, userAndGroups) ?
+                        elem : null);
+    }
+
+    private <T> Collection<T> removeUnauthorizedAndNullEntities(Collection<T> elems,
+                                                                EntityFilterFunction<T> filterFunc)
+            throws SchemaNotFoundException {
+        ArrayList<T> res = new ArrayList<>();
+        for(T elem : elems) {
+            try {
+                T newElem = filterFunc.filter(elem);
+                if(newElem != null) {
+                    res.add(newElem);
+                }
+            } catch (AuthorizationException e) { }
+        }
+
+        return res;
+    }
+
+    private void authorize(Authorizer.Resource resource, AccessType accessType, UserAndGroups userAndGroups)
+            throws AuthorizationException {
+
+        boolean isAuthorized = authorizer.authorize(resource, accessType, userAndGroups);
+
+        raiseAuthorizationExceptionIfNeeded(isAuthorized,
+                userAndGroups.getUser(),
+                accessType,
+                resource);
+    }
+
+    private String getPrimaryBranch(Collection<SchemaBranch> branches) {
+        return branches.stream().min((b1, b2) -> b1.getId() < b2.getId() ? -1 :
+                b1.getId() > b2.getId() ? 1 : 0).get().getName();
+    }
+
+    private void raiseAuthorizationExceptionIfNeeded(boolean isAuthorized,
+                                                     String user,
+                                                     AccessType permission,
+                                                     Authorizer.Resource resource) throws AuthorizationException {
+        if(!isAuthorized) {
+            throw new AuthorizationException(String.format(
+                    "User '%s' does not have [%s] permission on %s", user,
+                    permission.getName(), resource));
+        }
+    }
+
+    /// This method is needed only for the unit tests
```

Comment:
Can you annotate this method with **@VisibleForTesting**

Reply:
Done!